iroh-sync/src/store/fs/migrate_v1_v2.rs

```diff
@@ -0,0 +1,139 @@
+use std::path::{Path, PathBuf};
+
+use anyhow::Result;
+use redb::{MultimapTableHandle, TableHandle};
+use redb_v1::{ReadableMultimapTable, ReadableTable};
+use tempfile::NamedTempFile;
+use tracing::info;
+
+macro_rules! migrate_table {
+    ($rtx:expr, $wtx:expr, $old:expr, $new:expr) => {{
+        let old_table = $rtx.open_table($old)?;
+        let mut new_table = $wtx.open_table($new)?;
+        let name = $new.name();
+        let len = old_table.len()?;
+        info!("migrate {name} ({len} rows)..");
+        let ind = (len as usize / 1000) + 1;
+        for (i, entry) in old_table.iter()?.enumerate() {
+            let (key, value) = entry?;
+            let key = key.value();
+            let value = value.value();
+            if i > 0 && i % 100 == 0 {
+                info!("    {name} {i:>ind$}/{len}");
+            }
+            new_table.insert(key, value)?;
+        }
+        info!("migrate {name} done");
+    }};
+}
+
+macro_rules! migrate_multimap_table {
+    ($rtx:expr, $wtx:expr, $old:expr, $new:expr) => {{
+        let old_table = $rtx.open_multimap_table($old)?;
+        let mut new_table = $wtx.open_multimap_table($new)?;
+        let name = $new.name();
+        let len = old_table.len()?;
+        info!("migrate {name} ({len} rows)");
+        let ind = (len as usize / 1000) + 1;
+        for (i, entry) in old_table.iter()?.enumerate() {
+            let (key, values) = entry?;
+            let key = key.value();
+            if i > 0 && i % 100 == 0 {
+                info!("    {name} {i:>ind$}/{len}");
+            }
+            for value in values {
+                let value = value?;
+                new_table.insert(key, value.value())?;
+            }
+        }
+        info!("migrate {name} done");
+    }};
+}
+
+pub fn run(source: impl AsRef<Path>) -> Result<redb::Database> {
+    let source = source.as_ref();
+    // create the database to a tempfile
+    let target = NamedTempFile::new()?;
+    let target = target.into_temp_path();
+    info!("migrate {} to {}", source.display(), target.display());
+    let old_db = redb_v1::Database::open(source)?;
+    let new_db = redb::Database::create(&target)?;
+
+    let rtx = old_db.begin_read()?;
+    let wtx = new_db.begin_write()?;
+
+    migrate_table!(rtx, wtx, old::AUTHORS_TABLE, new::AUTHORS_TABLE);
```

Comment:
why not a function?

Reply:
It would need a lot of nested generics and I was a bit lazy. But can change I think.